File: java/src/com/alofmethbin/jadawin/styles/News.java
package com.alofmethbin.jadawin.styles;

import com.alofmethbin.jadawin.Article;
import com.alofmethbin.jadawin.Link;
import com.alofmethbin.jadawin.Page;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class News extends BaseStyle {
    private void findDatedArticles( Page page, List<Page> found) {
      if (page instanceof Link) {return;}
      Article article = (Article) page;
      
      if (article.hasDate() && article.isLeaf()) {
        found.add( article);
      }

      if (! article.isLeaf()) {
        for (Page child: article.children()) {
          findDatedArticles( child, found);
        }
      }
    }

    @Override
    public void prepare(Article article) {
        if ( article.hasDate() ) {
            article.error( "News style article must not be dated");
        }

        if ( article.hasChildren()) {
            article.error( "News style article must not have children");
        }

        List<Page> dated = new ArrayList<>();
        findDatedArticles( article, dated);
        sort( dated);
        
        for (int i = Math.min( 29, dated.size() - 1); i >= 0; i--) {
            article.addChild( dated.get(i));
        }
    }

    @Override
    public void sort( List<Page> children) {
        children.sort( new Comparator<>() {
            @Override
            public int compare( Page p0, Page p1) {
                return p1.getDate().compareTo( p0.getDate());
            }
        });
    }
}

File: ruby/styles/news.rb
require_relative '../link'

module Styles
  class News < Base
    def find_dated_articles( article, found)
      return if article.is_a?( Link)

      if article.date && article.leaf?
        found << article
      end

      unless article.leaf?
        article.children.each do |child|
          find_dated_articles( child, found)
        end
      end
    end

    def prepare( compiler, article, parents)
      if article.date
        article.error( 'News style article must not be dated')
        return
      end

      if article.children?
        article.error( 'News style article must not have children')
      end

      dated_articles = []
      find_dated_articles( parents[-1], dated_articles)
      dated_articles = dated_articles.uniq.sort_by do |dated|
        dated.date
      end

      if dated_articles.size > 30
        dated_articles = dated_articles[-30..-1]
      end

      dated_articles.reverse.each do |dated|
        title = dated.title + dated.date.strftime( ' (%b %Y)')
        article.add_child( Link.new( dated, dated, title))
      end
    end

   def sort( articles)
     articles.sort_by do |dated|
       - dated.date.to_i
     end
   end
  end
end
